pkg/aggregateeventrecorder/aggregateeventrecorder.go

```diff
@@ -0,0 +1,172 @@
+	StorageProviderID string    `json:"storageProviderId"` // The ID of the storage provider that served the retrieval content
+	TimeToFirstByte   float64   `json:"timeToFirstByte"`   // The time it took to receive the first byte in seconds
+	Bandwidth         float64   `json:"bandwidth"`         // The bandwidth of the retrieval in bytes per second
+	Success           bool      `json:"success"`           // Wether or not the retreival ended with a success event
+	StartTime         time.Time `json:"startTime"`         // The time the retrieval started
+	EndTime           time.Time `json:"endTime"`           // The time the retrieval ended
+}
+
+type batchedEvents struct {
+	Events []AggregatedEvent `json:"events"`
+}
+
+// NewAggregateEventRecorderSubsciber returns a RetrievalEventSubscriber that POSTs
+// a batch of aggregated retrieval events to an event recorder API endpoint
+func NewAggregateEventRecorderSubscriber(
+	ctx context.Context,
+	instanceID string,
+	endpointURL string,
+	endpointAuthorization string,
+	disableIndexerEvents bool,
+) types.RetrievalEventSubscriber {
+
+	incomingChan := make(chan types.RetrievalEvent)
+	postChan := make(chan []AggregatedEvent)
+	eventTempMap := make(map[string]tempData)
+
+	// Ingest data
+	go func() {
+		var batchedData []AggregatedEvent
+		var emptyGaurdChan chan []AggregatedEvent = nil
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			// read incoming data
+			case event := <-incomingChan:
+				id := event.RetrievalId().String()
+
+				switch event.Code() {
+				case types.StartedCode:
+					// Save the start time so we can calculate time to first byte
+					eventTempMap[id] = tempData{event.Time(), 0}
+
+				case types.FirstByteCode:
+					// Calculate time to first byte
+					tempData := eventTempMap[id]
+					tempData.ttfb = event.Time().Sub(tempData.startTime).Seconds()
+					eventTempMap[id] = tempData
+
+				case types.SuccessCode, types.FailedCode:
+					tempData := eventTempMap[id]
+
+					success := event.Code() == types.SuccessCode
+					spId := event.StorageProviderId().String()
+					endTime := event.Time()
+
+					var bytesPerSec float64
+					if success {
+						byts := event.(events.RetrievalEventSuccess).ReceivedSize()
+						duration := event.(events.RetrievalEventSuccess).Duration().Seconds()
+						bytesPerSec = float64(byts) / duration
+					}
+
+					aggregatedEvent := AggregatedEvent{
+						InstanceID:        instanceID,
+						RetrievalID:       id,
+						StorageProviderID: spId,
+						TimeToFirstByte:   tempData.ttfb,
+						Bandwidth:         bytesPerSec,
+						Success:           success,
+						StartTime:         tempData.startTime,
+						EndTime:           endTime,
+					}
+
+					batchedData = append(batchedData, aggregatedEvent)
+					if emptyGaurdChan == nil {
+						emptyGaurdChan = postChan // emptyGuardChan is no longer nil when we add to the batch
+					}
+				}
+			case emptyGaurdChan <- batchedData: // won't execute while emptyGaurdChan is nil
```

Comment:
interesting trick I don't think I've seen before, this is safe because we only touch it in this goroutine, right?

Reply:
This is the same trick currently being used in the current EventRecorder! I just renamed it to be clearer as to what it was doing.